File: backend/domain/approval/approval.go
// SPDX-FileCopyrightText: 2025 Mercedes-Benz Group AG and Mercedes-Benz AG
//
// SPDX-License-Identifier: Apache-2.0

package approval

import (
	"time"

	"github.com/eclipse-disuko/disuko/domain"
	"github.com/eclipse-disuko/disuko/domain/audit"
	"github.com/eclipse-disuko/disuko/domain/project/approvable"
	"github.com/eclipse-disuko/disuko/domain/project/components"
)

type ApproveState struct {
	State   StateInfo
	Updated *time.Time
}

type StateInfo string

const (
	Unset            StateInfo = ""
	Pending          StateInfo = "PENDING"
	Declined         StateInfo = "DECLINED"
	Approved         StateInfo = "APPROVED"
	SupplierApproved StateInfo = "SUPPLIER_APPROVED"
	CustomerApproved StateInfo = "CUSTOMER_APPROVED"
	Aborted          StateInfo = "ABORTED"
	GeneratingDocs   StateInfo = "GENERATING"
	GenerationFailed StateInfo = "GENERATION_FAILED"
)

func ParseStateInfo(state string) (valid bool, result StateInfo) {
	switch state {
	case string(Pending):
		valid, result = true, Pending
	case string(Declined):
		valid, result = true, Declined
	case string(Approved):
		valid, result = true, Approved
	case string(SupplierApproved):
		valid, result = true, SupplierApproved
	case string(CustomerApproved):
		valid, result = true, CustomerApproved
	case string(Aborted):
		valid, result = true, Aborted
	case string(GeneratingDocs):
		valid, result = true, GeneratingDocs
	case string(GenerationFailed):
		valid, result = true, GenerationFailed
	default:
		valid, result = false, Unset
	}
	return
}

type Approver int

const (
	Supplier1 Approver = 0
	Supplier2 Approver = 1
	Customer1 Approver = 2
	Customer2 Approver = 3
	None      Approver = 4
)

type ApprovalType string

const (
	TypeInternal     ApprovalType = "INTERNAL"
	TypePlausibility ApprovalType = "PLAUSIBILITY"
	TypeExternal     ApprovalType = "EXTERNAL"
)

type TaskMetaDocument struct {
	C1 bool
	C2 bool
	C3 bool
	C4 bool
	C5 bool
	C6 bool
}

func (d TaskMetaDocument) ToDto() DocumentFlagsDto {
	return DocumentFlagsDto{
		C1: d.C1,
		C2: d.C2,
		C3: d.C3,
		C4: d.C4,
		C5: d.C5,
		C6: d.C6,
	}
}

type ProjectApprovable struct {
	ProjectKey     string
	ProjectName    string
	ApprovableSPDX approvable.ApprovableSPDX
	// TODO: We don't need them anymore?
	CustomerDiffers bool
	SupplierDiffers bool
	ApprovableStats components.ComponentStats
	SpdxName        string
	SpdxTag         string
	SpdxUploaded    *time.Time
	IsSpdxRecent    bool

	Supplier           *string
	IsSpdxApprovable   bool
	HasProjectApproval bool
}

type Info struct {
	CompStats *components.ComponentStats
	Projects  []ProjectApprovable
}

type InternalApproval struct {
	ApproveStates    [4]ApproveState
	Approver         [4]string
	ApproveComments  [4]string
	DocVersion       int
	Aborted          bool
	Generating       bool
	GenerationFailed bool
}

type PlausibilityCheck struct {
	State          ApproveState
	ApproveComment string
	Approver       string
	Aborted        bool
}

type ExternalApproval struct {
	Vehicle bool
	Comment string
	State   StateInfo
}

type Approval struct {
	domain.ChildEntity `bson:"inline"`
	audit.Container    `bson:"inline"`

	ProjectGuid   string
	Creator       string
	Comment       string
	Info          Info
	DocumentFlags TaskMetaDocument

	Type         ApprovalType
	Internal     InternalApproval
	Plausibility PlausibilityCheck
	External     ExternalApproval
}

func (p *PlausibilityCheck) GetPendingApprover() string {
	if p.State.State != Pending {
		return ""
	}
	return p.Approver
}

func (p *PlausibilityCheck) IsActive() bool {
	return p.State.State == Pending && !p.Aborted
}

type PowerOfAttorneyType string

const (
	PowerOfAttorneyIV    PowerOfAttorneyType = "iV"
	PowerOfAttorneyPPA   PowerOfAttorneyType = "ppa"
	PowerOfAttorneyOther PowerOfAttorneyType = "other"
)

File: frontend/libs/portal/model/Approval.ts
// SPDX-FileCopyrightText: 2025 Mercedes-Benz Group AG and Mercedes-Benz AG
//
// SPDX-License-Identifier: Apache-2.0

import {DocumentDto} from '@disclosure-portal/model/Document';
import {ApprovableSPDXDto} from '@disclosure-portal/model/Project';
import {ComponentStats} from './VersionDetails';

export class ProjectApprovable {
  public projectKey = '';
  public projectName = '';
  public spdxname = '';
  public spdxtag = '';
  public spdxUploaded = '';
  public isSpdxRecent = '';
  public customerdiff = false;
  public supplierdiff = false;
  public stats: ComponentStats = {} as ComponentStats;
  public approvablespdx: ApprovableSPDXDto = {} as ApprovableSPDXDto;
  public supplier = '';
  public isSpdxApprovable = false;
  public hasProjectApproval = false;
}

export class ApprovableInfo {
  public stats: ComponentStats = {} as ComponentStats;
  public projects: ProjectApprovable[] = [];
  public hasDeniedDecisions = false;
}

export enum ApprovalStates {
  Pending = 'PENDING',
  Declined = 'DECLINED',
  Approved = 'APPROVED',
  SupplierApproved = 'SUPPLIER_APPROVED',
  CustomerApproved = 'CUSTOMER_APPROVED',
  Aborted = 'ABORTED',
  Generating = 'GENERATING',
  GenerationFailed = 'GENERATION_FAILED',
}

export enum ApproverRoles {
  Supplier1 = 0,
  Supplier2 = 1,
  Customer1 = 2,
  Customer2 = 3,
}

export class ApproveState {
  public updated = '';
  public state: ApprovalStates = ApprovalStates.Pending;
}

export class InternalApproval {
  public states: ApproveState[] = [];
  public approver: string[] = [];
  public approverFullName: string[] = [];
  public comments: string[] = [];
  public docVersion = 0;
  public aborted = false;
}

export class ExternalApproval {
  public vehicle = false;
  public state: ApprovalStates = ApprovalStates.Pending;
  public comment = '';
}

export class Plausibility {
  public state: ApproveState = {state: ApprovalStates.Pending, updated: ''};
  public comment = '';
  public approver = '';
  public approverFullName = '';
}

export enum ApprovalType {
  Internal = 'INTERNAL',
  Plausibility = 'PLAUSIBILITY',
  External = 'EXTERNAL',
}

export class DocumentFlags {
  public c1 = false;
  public c2 = false;
  public c3 = false;
  public c4 = false;
  public c5 = false;
  public c6 = false;
}

export class Approval {
  public key = '';
  public created = '';
  public updated = '';

  public projectKey = '';
  public creator = '';
  public creatorFullName = '';
  public comment = '';

  public info: ApprovableInfo = {} as ApprovableInfo;
  public documents: DocumentDto[] = [];
  public flags: DocumentFlags = {} as DocumentFlags;

  public type: ApprovalType = ApprovalType.Internal;

  public internal: InternalApproval = {} as InternalApproval;
  public plausibility: Plausibility = {} as Plausibility;
  public external: ExternalApproval = {} as ExternalApproval;

  public status: ApprovalStates = ApprovalStates.Pending;
}

export enum PowerOfAttorneyType {
  iV = 'iV',
  ppa = 'ppa',
  other = 'other',
}

export class ApprovalUpdate {
  public state = ApprovalStates.Pending;
  public comment = '';

  public powerOfAttorney!: PowerOfAttorneyType;
}

export class CreateApprovalResponse {
  public approvalGuid = '';
  public success = false;
  public jobKey = '';
}
